### DISCUSSION TITLE
Inheritance in DAOS

### DISCUSSION THREAD

**Discussion by ArthurValadares** [simolus3/drift#2866] [Q&A] [answered]
Hello everybody.
I would like to clarify a doubt regarding the possibility of inheritance in the context of DAOs.

During the modeling of any database, I realized the possibility of using inheritance so that different classes, tables, from the same group, would have the same set of columns. In order to do this I created a mixin.

In order to keep the database interaction code modular, I would create a DAO for each table. However, I realized that I would have to repeat the same block of code - with few changes - for each table, in order to share the same set of columns that it inherited from the Auditable mixin, which contains: 

- createdAt;
- updatedAt;
- deletedAt;

All DateTimes. The set of functions they share are to search based on the value equal, less than, greater than, greater than or equal, less than or equal in rows in which the values ​​are active or not.

The only element that changes in each of the functions is the table from which the search will be made.Is there any way to create another mixin, an abstract class or something like that so that, through inheritance, this repetition can be eliminated?

**Comment by simolus3** [comment]
> Is there any way to create another mixin, an abstract class or something like that so that, through inheritance, this repetition can be eliminated?

Sure. Let's say you had a common mixin defining these columns:

```dart
mixin WithDates on Table {
  DateTimeColumn get createdAt => ...;
  // ...
}

class Users extends Table with WithDates {}
```

That allows writing a common DAO for all of them:

```dart
abstract class BaseDao<D, T extends TableInfo<WithDates, D>> extends DatabaseAccessor {
  BaseDao(super.attachedDatabase);

  T get table;

  Future<D> testQuery() async {
    return await (select(table)..where((tbl) => tbl.createdAt.equals(whatever))).getSingle();
  }
}
```

A DAO for a particular table would extend that one, e.g. `class UsersDao extends BaseDao<User, $UsersTable>`.

**Reply by ArthurValadares** [reply]
Sorry, it must be my stupidity, but the generated code does not have a class that meets the T parameter type requirement.
Should I make any build changes?

**Reply by simolus3** [reply, marked as answer]
Not at all, the code I've posted was wrong. The first type parameter on `TableInfo` is the source table class that you've added to the database, the second one is the generated row class. So it should be `BaseDao<D, T extends TableInfo<WithDates, D>>` and then the `UsersDao` would extend `BaseoDao<User, $UsersTable>`.

**Reply by ArthurValadares** [reply]
Thank you very, very much.
